core/vote/vote_pool.go

```diff
@@ -0,0 +1,401 @@
+func (pool *VotePool) loop() {
+	for {
+		select {
+		// Handle ChainHeadEvent.
+		case ev := <-pool.chainHeadCh:
+			if ev.Block != nil {
+				latestBlockNumber := ev.Block.NumberU64()
+				pool.prune(latestBlockNumber)
+				pool.transferVotesFromFutureToCur(ev.Block.Header())
+			}
+		case <-pool.chainHeadSub.Err():
+			return
+
+		// Handle votes channel and put the vote into vote pool.
+		case vote := <-pool.votesCh:
+			pool.putIntoVotePool(vote)
+		}
+	}
+}
+
+func (pool *VotePool) PutVote(vote *types.VoteEnvelope) {
+	pool.votesCh <- vote
+}
+
+func (pool *VotePool) putIntoVotePool(vote *types.VoteEnvelope) bool {
+	targetNumber := vote.Data.TargetNumber
+	targetHash := vote.Data.TargetHash
+	header := pool.chain.CurrentBlock().Header()
+	headNumber := header.Number.Uint64()
+
+	voteData := &types.VoteData{
+		TargetNumber: targetNumber,
+		TargetHash:   targetHash,
+	}
+
+	var votes map[common.Hash]*VoteBox
+	var votesPq *votesPriorityQueue
+	isFutureVote := false
+
+	voteBlock := pool.chain.GetHeaderByHash(targetHash)
+	if voteBlock == nil {
+		votes = pool.futureVotes
+		votesPq = pool.futureVotesPq
+		isFutureVote = true
+	} else {
+		votes = pool.curVotes
+		votesPq = pool.curVotesPq
+	}
+
+	voteHash := vote.Hash()
+	if ok := pool.basicVerify(vote, headNumber, votes, isFutureVote, voteHash); !ok {
+		return false
+	}
+
+	if !isFutureVote {
+		// Verify if the vote comes from valid validators based on voteAddress (BLSPublicKey), only verify curVotes here, will verify futureVotes in transfer process.
+		if posa, ok := pool.engine.(consensus.PoSA); ok {
+			if !posa.VerifyVote(pool.chain, vote) {
+				return false
+			}
+		}
+		// Send vote for handler usage of broadcasting to peers.
+		voteEv := core.NewVoteEvent{Vote: vote}
+		pool.votesFeed.Send(voteEv)
+	}
+
+	pool.putVote(votes, votesPq, vote, voteData, voteHash, isFutureVote)
+
+	return true
+}
+
+func (pool *VotePool) SubscribeNewVoteEvent(ch chan<- core.NewVoteEvent) event.Subscription {
+	return pool.scope.Track(pool.votesFeed.Subscribe(ch))
+}
+
+func (pool *VotePool) putVote(m map[common.Hash]*VoteBox, votesPq *votesPriorityQueue, vote *types.VoteEnvelope, voteData *types.VoteData, voteHash common.Hash, isFutureVote bool) {
+	targetHash := vote.Data.TargetHash
+	targetNumber := vote.Data.TargetNumber
+
+	log.Info("The vote info to put is:", "voteBlockNumber", targetNumber, "voteBlockHash", targetHash)
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if _, ok := m[targetHash]; !ok {
+		// Push into votes priorityQueue if not exist in corresponding votes Map.
+		// To be noted: will not put into priorityQueue if exists in map to avoid duplicate element with the same voteData.
+		heap.Push(votesPq, voteData)
+		voteBox := &VoteBox{
+			blockNumber:  targetNumber,
+			voteMessages: make([]*types.VoteEnvelope, 0, maxCurVoteAmountPerBlock),
+		}
+		m[targetHash] = voteBox
+
+		if isFutureVote {
+			localFutureVotesPqGauge.Inc(1)
```

Comment:
Please use `Inc` as much as possible, use `Update(votesPq.Len())` instead. It can avoid the logic caused error.